db/revision_cache_lru.go

```diff
@@ -147,7 +147,10 @@ func (rc *LRURevisionCache) getFromCache(ctx context.Context, docID, revID strin
 		return DocumentRevision{}, nil
 	}
 
-	if value.invalid {
+	value.lock.RLock()
```

Comment:
This is the check that has performance implications, particularly on systems with a large number of cores, where acquiring a read lock can require scheduling work across cores.  We've seen this in our performance testing in the past, which is how we ended up with the design that attempts to keep the value immutable.  
As discussed, the invalidation flag breaks this immutability so we're going to need to change something.  I don't think switching back to acquiring a read lock on value is the ideal approach, as invalidation is an uncommon use case, and we'd be taking on that cost on every rev cache read.